draft-ietf-quic-transport.md

```diff
@@ -999,8 +999,10 @@ retired are considered active; any active connection ID can be used.
 An endpoint SHOULD ensure that its peer has a sufficient number of available and
 unused connection IDs.  Endpoints store received connection IDs for future use
 and advertise the number of connection IDs they are willing to store with the
-active_connection_id_limit transport parameter.  An endpoint SHOULD NOT provide
-more connection IDs than the peer's limit.
+active_connection_id_limit transport parameter.  An endpoint MUST NOT provide
+more connection IDs than the peer's limit.  An endpoint that receives more
+connection IDs than its advertised active_connection_id_limit MUST close the
+connection with an error of type CONNECTION_ID_LIMIT_ERROR.
```

Comment:
We're requiring other limit checks (e.g. for STREAM_LIMIT_ERROR and for FLOW_CONTROL_ERROR), so I'd prefer to keep this a requirement as well.

I don't think dropping should be regarded as acceptable behavior (although you can't prevent implementations from doing it), since a CID issued by an endpoint consumes state on that endpoint as well.

Reply:
I'm in agreement with @MikeBishop -- I would suggest changing this to a SHOULD error. This is different than the flow control errors, since an endpoint does not have to store all connection IDs that have been received.

Reply:
I might suggest having a design discussion in the issue (i.e. #3193) rather than here. There has been certain amount of discussion in that issue, and it seems to me that we have preference for MUST NOT.

Reply:
I'm not sure that I'm against the SHOULD, but I think the idea behind enforcing this was that an endpoint would indeed now need to store every CID it receives, and if it's not willing to store X amount, then it shouldn't have advertised a limit of X.

And even if you *still* didn't store them all, you technically don't need to store them to keep track of how many you've been issued, which you'll probably need in order to cope with getting Retire Prior To anyways.

Reply:
I'm ok with this being a MUST. I would suggest that you explicitly add a note saying that an endpoint is expected to store enough information all received CIDs to be able to generate this error. It might be redundant, but that's ok.

Reply:
I don't think this shhould be MUST. For every other such action, the paradigm is "the sender MUST NOT do this silly stuff, the responder MAY close with connection error"

Reply:
@huitema That's not correct:
> Endpoints MUST NOT exceed the limit set by their peer. An endpoint that receives a frame with a stream ID exceeding the limit it has sent MUST treat this as a connection error of type STREAM_LIMIT_ERROR 

and 
>  An endpoint MUST terminate a connection with a FLOW_CONTROL_ERROR error if it receives more data than the maximum data value that it has sent

Reply:
@marten-seemann The examples that you quote are cases where there is no reasonable way to continue the connection. I don't think that excess number of NEW CONNECTION ID is in the same class -- the receiving peer could also just drop them without retiring them.

Reply:
@huitema That doesn’t work well with Retire Prior To, since you still MUST retire that connection ID. So you need to track at least the sequence number of the CID.
It seems like you’re trying to build a way for a peer to (silently) revoke the limit it announced in its transport parameters. Building such a mechanism is possible (we’d need something like a MAX_CID frame), but we decided to not do this, mostly because we assumed that the number of requested connection IDs will be small (O(10)). If that argument still holds, I don’t think it’s too much to ask for to keep track of O(10) CID sequence numbers.